contracts/lib/IdentityBase.sol

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.16;
+pragma abicoder v2;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "../interfaces/IState.sol";
+import "../lib/ClaimBuilder.sol";
+import "../lib/OnChainIdentity.sol";
+
+// /**
+//  * @dev Contract managing onchain identity
+//  */
+contract IdentityBase is OwnableUpgradeable {
+    using OnChainIdentity for OnChainIdentity.Trees;
+    using OnChainIdentity for OnChainIdentity.IdentityData;
+
+    OnChainIdentity.IdentityData public identity;
+    OnChainIdentity.Trees internal treeRoots;
+    OnChainIdentity.LastTreeRoots public lastTreeRoots;
+
+    function getMaxSmtDepth() public pure virtual returns (uint256) {
+        return 40;
+    }
+
+    function initialize(address _stateContractAddr) public virtual initializer {
+        identity.initialize(_stateContractAddr, 
+            getMaxSmtDepth(),
+            treeRoots);
+
+        __Ownable_init();
+    }
+
+    /**
+     * @dev Add claim
+     * @param claim - claim data
+     */
+    function addClaim(uint256[8] memory claim) public virtual onlyOwner {
```

Comment:
It may be that many of such methods should have "external" but not "public" visibility and "calldata" but not "memory" arguments allocation. This saves some gas sometimes.
You may rut the **slither** static analysis on the code to prompt for such optimisations throughout all the code.

Reply:
replaced to calldata where possible